### PR TITLE
Fix Issue 17679 - SortedRange.contains should be deprecated in favor of the generic canFind

### PR DESCRIPTION
**Pull request by wilzbach**
See also: http://forum.dlang.org/post/sjwsaofnanozytrvwxpb@forum.dlang.org

### CODE REVIEW

File: std/range/package.d [hunk 1]
```diff
@@ -9727,6 +9730,21 @@ sorting relation.
...
+    assert(r.canFind(3));
+    assert(!r.canFind(32));
+    auto r1 = sort!"a > b"(a);
+    assert(r1.canFind(3));
+    assert(!r1.canFind(32));
+    assert(r1.release() == [ 64, 52, 42, 3, 2, 1 ]);
+}
+
+deprecated
+@safe unittest
```
Comment (wilzbach) [inline std/range/package.d]: I duplicated the relevant unittest, s.t. we continue to test `contains`

File: std/range/package.d [hunk 2]
```diff
@@ -9701,8 +9701,11 @@ Returns $(D true) if and only if $(D value) can be found in $(D
 range), which is assumed to be sorted. Performs $(BIGOH log(r.length))
 evaluations of $(D pred). See also STL's $(HTTP
 sgi.com/tech/stl/binary_search.html, binary_search).
- */
 
+$(RED Deprecated - please use $(REF canFind, std,algorithm.searching) instead).
+ */
+    // @@@DEPRECATED_2018-02@@@
+    deprecated("Please use std.algorithm.searching : canFind instead.")
```
Comment (wilzbach) [inline std/range/package.d]: @jmdavis - I hope I have done it correctly this time?
IIRC the stages were:
- mark in documentation as deprecated
- remove from documentation
- remove entirely

**Comment by bubnenkoff** [comment]
Should canFind be aliased/renamed to contains? It's more general naming.

**Comment by andralex** [comment]
The apparent duplication is intentional and useful. `canFind` means "can run linear find and find this value", whereas `contains` means "quickly rummage for this value" and must finish in logarithmic time.

**Comment by JackStouffer** [comment]
@andralex But canFind specializes for SortedRange, so there's no speed benefit to using one over the other.

**Comment by andralex** [comment]
That's an opportunistic optimization. Calling canFind means "I don't care whether this is linear, just find the thing". It's nice to have that use contains if available.

**Comment by wilzbach** [comment]
> @andralex But canFind specializes for SortedRange, so there's no speed benefit to using one over the other.

Yep copy/pasting from Bugzilla:

> Since https://github.com/dlang/phobos/pull/4907 (December 2016), find takes advantage of Sortedness:
> https://github.com/dlang/phobos/blob/57b8d2511ea37cd01665b9b663d8ba246144b1f9/std/algorithm/searching.d#L1509

**Comment by andralex** [comment]
@wilzbach the usefulness of the distinction is in generic code that wants to make sure fast searching is available

**Comment by wilzbach** [comment]
> the usefulness of the distinction is in generic code that wants to make sure fast searching is available

Isn't it the opposite for generic code? It should only care about the `canFind` operation. 
Btw I just looked at the Phobos code and the only time a function cares about whether sth. is of type `SortedRange` is `find`.
Of course, there was https://github.com/dlang/phobos/pull/3534 which proposed to do more specializations like: with `isSortedRange!(Range, pred)`, but we don't even have `isSortedRange` and afaict the purpose of the `SortedRange` was to do the optimizations in the function (e.g. `canFind`) a la:

```d
static if(isSortedRange!(Range, pred)
 // faster path
else
 // normal path
```

and not at the call site:

```d
static if (hasMember!(Range, "contains"))
    r.contains(42);
else
    r.canFind(42);
```

Anyway I don't care about this too strongly. I just found on the NG that people struggle with it and find it confusing.

**Comment by andralex** [comment]
Offering functions of distinct complexity is within the doctrine of generic programming; there's long established practice of such in the STL, viz. member vs. non-member `find`. The same applies to DbI at least for as long as there's no systematic way of getting the complexity of an algorithm by means of introspection. My BigO library may challenge that notion in the future, but as things stand reflecting complexity in the name of functions is robust established practice.

The Achilles's heel of checking for `SortedRange` is that not only sorted ranges have fast `contains` - a variety of hashed containers come to mind. Any range with fast searching should implement `contains`. That information is vital and could make the difference between working and practically impossible as anyone who's joined two tables would know.

**Comment by JackStouffer** [comment]
@wilzbach Since Andrei has a good argument in favor and there's been no activity on this, I'm closing this. Please reopen if you have a counter point.

**Pull request closed without merging**